server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/ObjectStorageBlobsDAO.java

```diff
@@ -79,53 +73,48 @@
         return SwiftKeystone3ObjectStorage.daoBuilder(testConfig);
     }
 
-    public CompletableFuture<ContainerName> createContainer(ContainerName name) {
-        return CompletableFuture.supplyAsync(() -> blobStore.createContainerInLocation(DEFAULT_LOCATION, name.value()))
-            .thenApply(created -> {
-                if (!created) {
-                    LOGGER.debug("{} already existed", name);
-                }
-                return name;
-            });
+    public Mono<ContainerName> createContainer(ContainerName name) {
+        return Mono.fromCallable(() -> blobStore.createContainerInLocation(DEFAULT_LOCATION, name.value()))
+            .filter(created -> created == false)
+            .doOnNext(ignored -> LOGGER.debug("{} already existed", name))
+            .thenReturn(name);
     }
 
     @Override
-    public CompletableFuture<BlobId> save(byte[] data) {
+    public Mono<BlobId> save(byte[] data) {
         return save(new ByteArrayInputStream(data));
     }
 
     @Override
-    public CompletableFuture<BlobId> save(InputStream data) {
+    public Mono<BlobId> save(InputStream data) {
         Preconditions.checkNotNull(data);
 
         BlobId tmpId = blobIdFactory.randomId();
         return save(data, tmpId)
-            .thenCompose(id -> updateBlobId(tmpId, id));
+            .flatMap(id -> updateBlobId(tmpId, id));
     }
 
-    private CompletableFuture<BlobId> updateBlobId(BlobId from, BlobId to) {
+    private Mono<BlobId> updateBlobId(BlobId from, BlobId to) {
         String containerName = this.containerName.value();
-        return CompletableFuture
-            .supplyAsync(() -> blobStore.copyBlob(containerName, from.asString(), containerName, to.asString(), CopyOptions.NONE), executor)
-            .thenAcceptAsync(any -> blobStore.removeBlob(containerName, from.asString()))
-            .thenApply(any -> to);
+        return Mono
+            .fromCallable(() -> blobStore.copyBlob(containerName, from.asString(), containerName, to.asString(), CopyOptions.NONE))
```

Comment:
Can't we benefit from specifying the elastic scheduler directly in this class??

Reply:
Shouldn't we `cache()` ?

What should happen when I do:

```
Mono<Void> saveOp = blobStore.save(...);
saveOp.block();
saveOp.block();
```

One time, or two time?

Can we add the tests in the BlobStore contract?

Reply:
> What should happen when I do:

> ```Mono<Void> saveOp = blobStore.save(...);
> saveOp.block();
> saveOp.block();
> ```

When you receive a Mono, you expect it to be runnable (you can subscribe to it to trigger the run).
There's nothing special in this code, it's the same we write for weeks.

Quoting Mono documentation:

> Subscribe to this Mono and block indefinitely until a next signal is received.

So if you call `block` twice, it run twice. Mono is like a Supplier instance.

What would you expect from this test ?